### PR TITLE
Robot visualization

### PR DESCRIPTION
**Pull request by jontromanab**
This commit targets several improvements mostly on base_placement

- #16 Showing robot model as visualization method for output
- #24 Output of base placement fixed

- The base placement methods are modified and corrected
- New method verticalrobotmodel has been introduced, which considers the base of the full robot model, not only the arm base.
- New method userintuition has been introduced, where the user can select the base positions and can compare results with algorithms
- 2d reachability map for robot base
- A new scoring system has been introduced which can measure the efficency of the solutions calculated
- Get the best pose for base placement for the task
- The base placement plugin now launches with moveit_pkg from where it takes the robot model and other features
- All the algorithms are tested with motoman_mh5, UR5 and abb_irb2400 (all the ikfast files and map files are provided)
- Base placement is tested on a mobile manipulator (husky_UR5) in gazebo

screenshot from 2017-02-21 11 43 09

screenshot from 2017-02-21 11 54 32

### CODE REVIEW

**Comment by jontromanab** [comment]
@akgoins have you checked this PR. Please let me know if you have some confusions. Lets work on this and merge this. I will modify the documentation after you merge it.

**Comment by akgoins** [comment]
I have looked at and and tried to run it.  I was a bit confused on the vertical robot method and would like some directions on how to use it and what the expected output should look like.  I have not done an in depth review but will do that soon.  I recently upgraded my work PC to 16.04 and I'm having some build issues which is keeping me from doing further testing.  I will either have to resolve the build issues or use a different PC to continue testing the code.

**Comment by jontromanab** [comment]
@akgoins Merging this PR is taking unusual time. Is there any problem? Can you please specify some parts where it is getting confusing ? I can specify with some screenshots and instructions. Please let me know.

File: base_placement_plugin/src/create_marker.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    Eigen::Affine3d tf_first_link_to_root = tf_root_to_first_link.inverse();//inverse of the transform
+
+    Eigen::Affine3d base_tf;
+    tf::poseMsgToEigen(base_pose, base_tf);//base pose in tf
+    visualization_msgs::InteractiveMarker iMarker; //new interactive marker
+    iMarker.header.frame_id = "base_link";
+    iMarker.pose = base_pose;//the pose of the imarker is the pose of the base
+    iMarker.scale = 0.2;
+    std::string name = "robot_model";
+    std::string description = "robot model";
```
Comment (akgoins) [inline base_placement_plugin/src/create_marker.cpp]: Are the strings `name` and `description` ever used anywhere?

File: base_placement_plugin/src/create_marker.cpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    iMarker.name = name+boost::lexical_cast<std::string>(i);
+    iMarker.description = description+boost::lexical_cast<std::string>(i);//each imarker has different name and description
+    visualization_msgs::InteractiveMarkerControl robotModelControl;
+    for(std::size_t j=0; j<full_link_markers.markers.size();j++)
+    {
+      full_link_markers.markers[j].header.frame_id = "base_link";
+      full_link_markers.markers[j].type = visualization_msgs::Marker::MESH_RESOURCE;
+      full_link_markers.markers[j].mesh_use_embedded_materials = true;
+      full_link_markers.markers[j].id = i*(j+5);
```
Comment (akgoins) [inline base_placement_plugin/src/create_marker.cpp]: What is the +5 for in this line?  Will this be valid for all robot configurations?

File: base_placement_plugin/src/create_marker.cpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+      if(no_joint_solution)
+      {
+        full_link_markers.markers[j].color.r = 1.0;
+        full_link_markers.markers[j].color.g = 0.0;
+        full_link_markers.markers[j].color.b = 0.0;
+        full_link_markers.markers[j].color.a = 0.2;
+      }
+      else
+      {
+        if(j==1 || j==2 || j==5 || j==6 || j==9)
```
Comment (akgoins) [inline base_placement_plugin/src/create_marker.cpp]: Again, what are these numbers?  Will they work for any robot models or only 6 DOF?

File: base_placement_plugin/src/create_marker.cpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    double b = unifRand();
+    for(std::size_t j=0; j<full_link_markers.markers.size();j++)
+    {
+      full_link_markers.markers[j].header.frame_id = "base_link";
+      full_link_markers.markers[j].type = visualization_msgs::Marker::MESH_RESOURCE;
+      full_link_markers.markers[j].mesh_use_embedded_materials = true;
+      full_link_markers.markers[j].id = i*(j+5);
+      full_link_markers.markers[j].header.stamp = ros::Time::now();
+      full_link_markers.markers[j].ns = "robot_links";
+      full_link_markers.markers[j].lifetime = ros::Duration(40.0);
```
Comment (akgoins) [inline base_placement_plugin/src/create_marker.cpp]: A lot of this code looks like it is replicated from the above function `makeArmMarker()`.  If there is enough reused code, it needs to be separated into its own function so we don't have two sections of code doing the same thing.

File: base_placement_plugin/src/create_marker.cpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    Eigen::Affine3d tf_first_link_to_root = tf_root_to_first_link.inverse();//inverse of the transform
+    Eigen::Affine3d base_tf;
+    tf::poseMsgToEigen(base_pose, base_tf);//base pose in tf
+
+    visualization_msgs::InteractiveMarker iMarker; //new interactive marker
+    iMarker.header.frame_id = "base_link";
+    iMarker.pose = base_pose;//the pose of the imarker is the pose of the base
+    iMarker.scale = 0.2;
+    std::string name = "robot_model";
+    std::string description = "robot model";
```
Comment (akgoins) [inline base_placement_plugin/src/create_marker.cpp]: Unused variables?

File: base_placement_plugin/src/place_base.cpp [hunk 6]
```diff
@@ -110,7 +182,184 @@ void PlaceBase::setBasePlaceParams(int base_loc_size, int high_score_sp)
...
+  std::vector<int>::const_iterator it;
+  it = max_element(poseCount.begin(), poseCount.end());
+  int max_number = *it;
+  it = min_element(poseCount.begin(), poseCount.end());
+  int min_number = *it;
+  if(reduce_D)
+  {
+    for (std::multimap< std::vector< double >, std::vector< double > >::iterator it = basePoses.begin(); it != basePoses.end();++it)
+    {
+      if((it->first)[2] < 0.051 && (it->first)[2] > -0.051)
```
Comment (akgoins) [inline base_placement_plugin/src/place_base.cpp]: What is the 0.051 for?  Are you just trying to check to see if it is near zero?  Does this value correspond to a certain angle?

**Comment by jontromanab** [comment]
@akgoins Thanks for reviewing. I am currently in Barcelona for a collaborative project visit. As not being available to my system right now, it is hard for me to discuss further about the errors. I will try to make everything work here, and will come back to you as soon as possible.

File: base_placement_plugin/src/create_marker.cpp (hunk 4 above)
Comment (jontromanab) [inline base_placement_plugin/src/create_marker.cpp]: @akgoins  I have created a new function. So now the create_marker is pretty modular. Please review the PR with the new changes.

File: base_placement_plugin/src/place_base.cpp (hunk 6 above)
Comment (jontromanab) [inline base_placement_plugin/src/place_base.cpp]: Yes. As we are disecting the map to find base placement locations for the whole robot near the floor, we are considering this range.

**Comment by jontromanab** [comment]
@akgoins I have also resolved the conflicts. Please have a look and merge this. We can discuss about the theories in the mail. When we agree, I will create a video and update the wiki.

**Comment by akgoins** [comment]
+1.  I found a few odd behaviors, but I will add those as issues so we can get this PR merged.

**Pull request merged**